### PR TITLE
[sil-combine] Fix cast optimizer based optimizations for ownership.

### PR DESCRIPTION
**Pull request by gottesmm**
The cast optimizer itself was updated pretty much, but I needed to update some helpers used to feed the cast optimizer.

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test source compatibility

**Comment by gottesmm** [comment]
Passed full macOS tests but failed on lldb tests going to smoke test

File: lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp [hunk 1]
```diff
@@ -593,20 +588,36 @@ visitUnconditionalCheckedCastAddrInst(UnconditionalCheckedCastAddrInst *UCCAI) {
...
+    {
+      auto *valInsertPt = val->getDefiningInsertionPoint();
+      if (!valInsertPt)
+        return nullptr;
+      auto valInsertPtIter = valInsertPt->getIterator();
+      // If our value is defined by an instruction (not an argument), we want to
+      // insert the copy after that. Otherwise, we have an argument and we want
+      // to insert the copy right at the beginning of the block.
+      if (val->getDefiningInstruction())
+        valInsertPtIter = std::next(valInsertPtIter);
```
Comment (atrick) [inline lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp:605]: Can you please just replace all this interator junk with `getNextInstruction`?

File: lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp [hunk 2]
```diff
@@ -593,20 +588,36 @@ visitUnconditionalCheckedCastAddrInst(UnconditionalCheckedCastAddrInst *UCCAI) {
...
-  if (canBeUsedAsCastDestination(val, UCCAI, DA)) {
-    SILBuilderContext builderCtx(Builder.getModule(), Builder.getTrackingList());
-    SILBuilderWithScope builder(UCCAI, builderCtx);
-    SILLocation loc = UCCAI->getLoc();
-    builder.createRetainValue(loc, val, builder.getDefaultAtomicity());
-    builder.createDestroyAddr(loc, UCCAI->getSrc());
-    builder.createStore(loc, val, UCCAI->getDest(),
-                        StoreOwnershipQualifier::Unqualified);
-    return eraseInstFromFunction(*UCCAI);
+  SILValue val = getConcreteValueOfExistentialBoxAddr(uccai->getSrc(), uccai);
```
Comment (atrick) [inline lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp:591]: The comment above is out-of-date now

File: lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp [hunk 3]
```diff
@@ -808,11 +810,27 @@ visitCheckedCastAddrBranchInst(CheckedCastAddrBranchInst *CCABI) {
...
+      auto *valInsertPt = val->getDefiningInsertionPoint();
+      if (!valInsertPt)
+        return nullptr;
+      auto valInsertPtIter = valInsertPt->getIterator();
+      // If our value is defined by an instruction (not an argument), we want to
+      // insert the copy after that. Otherwise, we have an argument and we want
+      // to insert the copy right at the beginning of the block.
+      if (val->getDefiningInstruction())
+        valInsertPtIter = std::next(valInsertPtIter);
+      SILBuilderWithScope builder(valInsertPtIter, Builder);
```
Comment (atrick) [inline lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp:828]: Same issue as above. Fix the comment and use `getNextInstruction`

File: lib/SILOptimizer/Utils/InstOptUtils.cpp [hunk 4]
```diff
@@ -701,41 +701,58 @@ SILValue swift::
 getConcreteValueOfExistentialBox(AllocExistentialBoxInst *existentialBox,
                                   SILInstruction *ignoreUser) {
   StoreInst *singleStore = nullptr;
-  for (Operand *use : getNonDebugUses(existentialBox)) {
+  SmallVector<Operand *, 32> worklist;
+  SmallPtrSet<Operand *, 8> addedToWorklistPreviously;
```
Comment (atrick) [inline lib/SILOptimizer/Utils/InstOptUtils.cpp:705]: We use `SmallSetVector` everywhere else for this pattern. If it's worth it to do this then we should add a thin `SmallPtrSetVector` wrapper and use that everywhere instead

File: lib/SILOptimizer/Utils/InstOptUtils.cpp [hunk 5]
```diff
@@ -753,29 +770,42 @@ getConcreteValueOfExistentialBoxAddr(SILValue addr, SILInstruction *ignoreUser)
...
-        break;
-      }
-      default:
-        if (stackUser != ignoreUser)
+    case SILInstructionKind::DestroyAddrInst: {
+      // Make sure the destroy_addr is the instruction before one of our
+      // dealloc_stack insts and is directly on the stack location.
+      auto next = std::next(stackUser->getIterator());
+      if (auto *dsi = dyn_cast<DeallocStackInst>(next))
+        if (dsi->getOperand() != stackLoc)
```
Comment (atrick) [inline lib/SILOptimizer/Utils/InstOptUtils.cpp:778]: This `destroy_addr` check is new, super-specific, brittle, and bizarre. Can you at least explain in the comment why you need it.

**Review by atrick — COMMENTED** [review]
Follow-up comments below

File: lib/SILOptimizer/Utils/InstOptUtils.cpp (hunk 4 above)
Comment (gottesmm) [reply lib/SILOptimizer/Utils/InstOptUtils.cpp:705]: I think SmallSetVector already uses a SmallPtrSet when it has a pointer operand. So it should just work. I think I just wrote this quickly.

File: lib/SILOptimizer/Utils/InstOptUtils.cpp [hunk 6]
```diff
@@ -701,41 +701,58 @@ SILValue swift::
...
+    case SILInstructionKind::DestroyValueInst:
+    case SILInstructionKind::EndBorrowInst:
+      break;
+    case SILInstructionKind::CopyValueInst:
+    case SILInstructionKind::BeginBorrowInst:
+      // Look through copy_value, begin_borrow
+      for (SILValue result : user->getResults())
+        for (auto *transitiveUse : result->getUses())
+          if (!addedToWorklistPreviously.insert(transitiveUse).second)
+            worklist.push_back(use);
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/InstOptUtils.cpp:726]: Was re-reading through this code fixing what you were talking about @atrick and found this thinko. I think this should be transitiveUse and I typed.

Comment (atrick) [reply lib/SILOptimizer/Utils/InstOptUtils.cpp:726]: oops

Comment (gottesmm) [reply lib/SILOptimizer/Utils/InstOptUtils.cpp:726]: yea... I'm not happy about it. But at least we caught it early!

**Pull request merged**